[BUGFIX][MON-PIX] Ne pas rediriger un utilisateur GAR sur la page de connexion (lors de la re-connexion, sur navigateur SAFARI) (PIX-6131)

igorissen opened the pull request:
## :christmas_tree: Problème

Sur le navigateur Safari (MacOS ou iOS), des élèves sont obligés d'effectuer plusieurs tentatives de connexion pour arriver sur la page d'accueil de Pix. Lorsque la connexion à Pix n'a pas abouti, les élèves arrivent sur la page de connexion de Pix, ce qui ne doit jamais arriver pour les élèves.

Après investigation, nous avons remarqué que le processus de connexion, en passant par le GAR, n'était pas adapté ce qui entrainait parfois une boucle de connexion et déconnexion (3 tentatives). Lors de cette boucle, une erreur survient et entraine la redirection de l'utilisateur vers la page de connexion de Pix.

La gestion des paramètres de type `query` par Ember lors d'un rafraichissement de page, ainsi que la gestion des sessions Ember Simple Auth après chargement de l'application sont la cause de cette boucle.

Nous avons également remarqué que le paramètre `user-id` n'était pas utilisé lors du processus de connexion sur Pix.

## :gift: Proposition

- Créer une route dédiée au processus de connexion via le GAR (**/connexion/saml**), ce qui permet de tester plus facilement cette connexion (pour un élève qui a déjà un compte)
- Utiliser un hash (`/connexion/saml#<access_token>`) dans l'URL au lieu du query (`/connexion/saml?token=<access_token>`)
- Retirer le paramètre `user-id` qui n'est pas utile
- Ajout d'un initializer pour supprimer les données d'Ember Simple Auth avant le chargement de l'application, si l'URL correspond à la connexion via le GAR
- Créer un authenticator dédié au GAR (`authenticator:gar`)
- Retirer le code qui gérait la connexion d'un élève ayant un compte Pix du service `session` (car la gestion de la connexion d'un élève se fait désormais par la nouvelle route `/connexion/saml` )

## :star2: Remarques

⚠️ Dû au faite que l'on fourni le token d'accès dans l'URL, nous pouvons retrouver ce token dans les logs. Il faudrait passer par une génération d'un code qui serait échangé par la suite pour récupérer le token.

## :santa: Pour tester

Ces 2 étapes sont a effectuer en priorité sur Safari (MacOS et iOS - tablette et/ou smartphone).

##### 1ère étape

- Se connecter via le GAR avec un élève **n'ayant pas** de compte chez Pix **(Prénom: Prenom, Nom: Nom)**
- Saisir le code `SCOBADGE1` pour accéder au parcours
- Commencer le parcours
- Finaliser la création du compte de l'élève en fournissant sa date de naissance **(Date de naissance: 09/09/2009)**
- Ignorer le didactitiel
- Quitter le parcours
- Constatez que l'utilisateur se retrouve sur la page d'accueil de Pix (`/accueil`)
- Se déconnecter
- Constatez que l'utilisateur se retrouve sur la page `/nonconnecte`

##### 2e étape

- Se connecter via le GAR avec le même élève **(Prénom: Prenom, Nom: Nom)**
- Constatez que l'utilisateur se retrouve sur la page d'accueil de Pix (`/accueil`)
- Se déconnecter
- Constatez que l'utilisateur se retrouve sur la page `/nonconnecte`

igorissen committed:
feat(mon-pix): add session instance initializer

This initializer will help reset EmberSimpleAuth current session when a user comes from a GAR authentication process

Co-authored-by: Anne-G <ang.schall@gmail.com>

pix-service commented:
I'm deploying this PR to these urls:

- App (.fr): https://app-pr5182.review.pix.fr
- App (.org): https://app-pr5182.review.pix.org
- Orga: https://orga-pr5182.review.pix.fr
- Certif: https://certif-pr5182.review.pix.fr
- Admin: https://admin-pr5182.review.pix.fr
- API: https://api-pr5182.review.pix.fr/api/

 Please check it out!

igorissen committed:
feat(mon-pix): add GAR authenticator

Co-authored-by: Anne-G <ang.schall@gmail.com>

igorissen committed:
feat(mon-pix): add new getter for pix window

Co-authored-by: Anne-G <ang.schall@gmail.com>

igorissen committed:
feat(mon-pix): add login-gar route

Co-authored-by: Anne-G <ang.schall@gmail.com>

igorissen committed:
feat(api): update redirect URI

Co-authored-by: Anne-G <ang.schall@gmail.com>

igorissen committed:
feat(mon-pix): handle GAR authentication with login-gar route

this task is no more handled by the session service by the _checkForURLAuthentication function

Co-authored-by: Anne-G <ang.schall@gmail.com>

igorissen committed:
fix(e2e): update commands

Co-authored-by: Anne-G <ang.schall@gmail.com>

Anne-Gaelle-S commented on mon-pix/app/authenticators/gar.js:
```diff
@@ -0,0 +1,21 @@
+import ENV from 'mon-pix/config/environment';
+import { decodeToken } from 'mon-pix/helpers/jwt';
+
+export default OAuth2PasswordGrant.extend({
```
**remarque**
Enlever le `OAuth2PasswordGrant` pour utilise le `baseAuthenticator` ?
Vu qu'on ne fait pas de revoke de token

Anne-Gaelle-S commented on mon-pix/tests/unit/authenticators/gar_test.js:
```diff
@@ -0,0 +1,20 @@
+describe('Unit | Authenticator | gar', function () {
+  setupTest();
+
+  it('should have token revocation endpoints', function () {
```
**suggestion**
Enlever ce test pour faire un test sur `authenticate` voir `mon-pix/tests/unit/authenticators/oidc_test.js`

Anne-Gaelle-S commented on mon-pix/tests/unit/routes/authentication/login-saml_test.js:
```diff
@@ -0,0 +1,31 @@
+    context('when a token is set as an hash of an url', function () {
+      it('should authenticate the user', async function () {
+        // given
+        const token = 'token';
```
**remarque**
mettre un autre texte que "token" à minima ? (symbol or not ?)

igorissen replied:
Malheureusement utiliser Symbol n'est pas possible à cause de la ligne 21 `.returns('#token')`.

Anne-Gaelle-S commented on api/tests/acceptance/application/saml-controller_test.js:
```diff
@@ -237,7 +237,7 @@ describe('Acceptance | Controller | saml-controller', function () {
       // then
       expect(response.statusCode).to.equal(302);
-      expect(response.headers.location).to.match(/^\/\?token=[-_a-zA-Z0-9.]+&user-id=[0-9]+$/);
+      expect(response.headers.location).to.match(/^\/connexion\/saml#[-_a-zA-Z0-9.]+$/);
```
MENSOOOONGE

igorissen committed:
fix(mon-pix): fix first disconnection problem

Co-authored-by: Anne-G <ang.schall@gmail.com>

clemlatz commented on mon-pix/app/instance-initializers/session.js:
```diff
@@ -0,0 +1,15 @@
+
+export function initialize(/* applicationInstance */) {
+  const currentURL = PixWindow.getLocationHref();
+  const isGarAuthenticationURL = currentURL.match(/\/connexion\/saml/i)?.length === 1;
```
**suggestion**
Est-ce que ce problème est propre à SAML ou au GAR ? Dans les deux cas, je propose d'être cohérent et de ne parler que de l'un ou de l'autre dans cette PR (idéalement on n'aurait que du SAML comme pour la brique OIDC) :

```suggestion
  const isSamlAuthenticationURL = currentURL.match(/\/connexion\/saml/i)?.length === 1;
```

ou

```suggestion
  const isGarAuthenticationURL = currentURL.match(/\/connexion\/gar/i)?.length === 1;
```

igorissen replied:
Je vais plutôt renommer l'URL en `/gar` au lieu de `/saml`. Si un jour on a un 2e partenaire avec une authentification de type SAML on repassera sur cette partie.

clemlatz commented on mon-pix/tests/unit/instance-initializers/session-test.js:
```diff
@@ -0,0 +1,58 @@
+  });
+
+  context('when a session exists', function () {
+    context('when a user finalize a GAR authentication process', function () {
```
**nitpick**

```suggestion
    context('when a user finalizes a GAR authentication process', function () {
```

clemlatz commented on mon-pix/app/authenticators/gar.js:
```diff
@@ -0,0 +1,18 @@
+import RSVP from 'rsvp';
+import { decodeToken } from 'mon-pix/helpers/jwt';
+
+export default class GarAuthenticator extends BaseAuthenticator {
```
**question**

Idem ici, est-ce que ce serait pertinent d'avoir un `SamlAuthenticator` ?

igorissen replied:
Lorsque nous aurons un autre partenaire avec une authentification de type SAML, oui ça aurait du sens de mettre SamlAuthenticator.

clemlatz commented:
✅ Test fonctionnel OK sur Safari (macOS)
